clients/venice-client/src/main/java/com/linkedin/venice/fastclient/transport/GrpcTransportClient.java

```diff
@@ -0,0 +1,191 @@
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+
+public class GrpcTransportClient extends InternalTransportClient {
+  private static final Logger LOGGER = LogManager.getLogger(GrpcTransportClient.class);
+  private final TransportClient r2TransportClient; // used for metadata requests
+
+  // field has public visibility for unit testing
+  public final Map<String, ManagedChannel> serverGrpcChannels;
+  final Map<String, String> veniceAddressToGrpcAddress;
+
+  public GrpcTransportClient(Map<String, String> nettyServerToGrpc, TransportClient r2TransportClient) {
+    serverGrpcChannels = new HashMap<>();
+    veniceAddressToGrpcAddress = new HashMap<>();
+    this.r2TransportClient = r2TransportClient;
+
+    for (Map.Entry<String, String> entry: nettyServerToGrpc.entrySet()) {
+      String nettyServer = entry.getKey().split(":")[0];
+      String grpcServer = nettyServer + ":" + entry.getValue();
+
+      veniceAddressToGrpcAddress.put(entry.getKey(), grpcServer);
+    }
+  }
+
+  public ManagedChannel getChannel(String requestPath) {
+    String portKey = requestPath.split("/")[2];
+
+    if (!veniceAddressToGrpcAddress.containsKey(portKey)) {
+      throw new VeniceException("No grpc server found for port: " + portKey);
+    }
+
+    return serverGrpcChannels.computeIfAbsent(
+        portKey,
+        k -> ManagedChannelBuilder.forTarget(veniceAddressToGrpcAddress.get(k)).usePlaintext().build());
+  }
+
+  @Override
+  public CompletableFuture<TransportClientResponse> get(String requestPath, Map<String, String> headers) {
+    String[] requestParts = RequestHelper.getRequestParts(requestPath);
+
+    if (!requestParts[1].equals("storage")) {
+      LOGGER.debug("Requesting metadata from R2 client");
+      return r2TransportClient.get(requestPath, headers);
+    }
+
+    // Lazily create channel for each venice server
+    ManagedChannel channel = getChannel(requestPath);
+
+    VeniceClientRequest request = VeniceClientRequest.newBuilder()
+        .setStoreName(requestParts[2])
+        .setResourceName(requestParts[2])
+        .setPartition(Integer.parseInt(requestParts[3]))
+        .setKeyString(requestParts[4])
+        .build();
+
+    VeniceReadServiceGrpc.VeniceReadServiceStub clientStub = VeniceReadServiceGrpc.newStub(channel);
+    GrpcTransportClientCallback callback = new GrpcTransportClientCallback(clientStub, request);
+
+    return callback.get();
+  }
+
+  @Override
+  public CompletableFuture<TransportClientResponse> post(
+      String requestPath,
+      Map<String, String> headers,
+      byte[] requestBody) {
+    String[] requestParts = RequestHelper.getRequestParts(requestPath);
+
+    ManagedChannel channel = getChannel(requestPath);
+    VeniceClientRequest request = VeniceClientRequest.newBuilder()
+        .setStoreName(requestParts[2])
```

Comment:
`StoreName` is typically different from `ResourceName`, so why we initialize them with the same value here?

Reply:
Yes, sorry. We do not need `StoreName` when sending the request to the server.